File: problems/0078/solution.cpp
#include <vector>
using namespace std;

class Solution {
 private:
  void backtrack(vector<int>& nums, int startIndex, vector<int>& path,
                 vector<vector<int>>& res) {
    res.push_back(path);

    for (int i = startIndex; i < nums.size(); i++) {
      path.push_back(nums[i]);
      backtrack(nums, i + 1, path, res);
      path.pop_back();
    }
  }

 public:
  vector<vector<int>> subsets(vector<int>& nums) {
    vector<int> path;
    vector<vector<int>> res;
    backtrack(nums, 0, path, res);
    return res;
  }
};

File: problems/0078/solution.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        path = []
        res = []

        def backtrack(startIndex: int):
            res.append(path[:])

            for i in range(startIndex, len(nums)):
                path.append(nums[i])
                backtrack(i + 1)
                path.pop()

        backtrack(0)
        return res
